---
title: "Gasperini (2019) data documentation"
author: "Gene Katsevich"
date: '2024-04-06'
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message = FALSE, warning = FALSE}
library(sceptre)
library(dplyr)
library(conflicted)
conflict_prefer("filter", "dplyr")
```

# Overview

This repository contains code to import and process the [Gasperini 2019](https://pubmed.ncbi.nlm.nih.gov/30612741/) data. Gasperini et al. developed a high MOI, single-cell CRISPR screen assay to map putative enhancers at genome-wide scale in a population of \>200,000 K562 cells.

The `gasperini-2019-v3` directory structure is as follows:

```
└── at-scale
    ├── intermediate
    ├── processed
    │   ├── grna.odm
    │   ├── response.odm
    │   ├── sceptre_object.rds
    │   └── sceptre_object_in_memory.rds
    └── raw
```

The contents of the `raw` and `intermediate` directories are suppressed, as they are unimportant. The `grna.odm`, `response.odm`, and `sceptre_object.rds` files are the three pieces of an ondisc-backed `sceptre_object`. The `sceptre_object_mem.rds` object is an in-memory `sceptre_object` containing the same information. It requires about 2GB on disk and 8GB in memory. Let us load the disk-backed `sceptre_object`:

```{r, message = FALSE}
gasp_offsite <- paste0(.get_config_path("LOCAL_GASPERINI_2019_V3_DATA_DIR"), "at-scale/")
processed_data_dir <- paste0(gasp_offsite, "processed/")
sceptre_object <- read_ondisc_backed_sceptre_object(
  sceptre_object_fp = paste0(processed_data_dir, "sceptre_object.rds"),
  response_odm_file_fp = paste0(processed_data_dir, "response.odm"),
  grna_odm_file_fp = paste0(processed_data_dir, "grna.odm")
)
```

We can print the `sceptre_object` for more information:
```{r}
sceptre_object
```

Let us take a look at the cell covariates:
```{r}
sceptre_object |> get_cell_covariates() |> head()
```

We can see that there are two batches in these data:
```{r}
sceptre_object |> get_cell_covariates() |> count(prep_batch)
```

Let us take a look at the gRNA-target data frame:
```{r}
sceptre_object@grna_target_data_frame
```

The gRNA id is the gRNA barcode. Different gRNAs have different target types.

Some target genes:
```{r}
sceptre_object@grna_target_data_frame |>
  filter(grepl("ENSG", grna_target))
```

Some target enhancers:
```{r}
sceptre_object@grna_target_data_frame |>
  filter(grepl("chr", grna_target))
```

Some are non-targeting:
```{r}
sceptre_object@grna_target_data_frame |>
  filter(grna_target == "non-targeting")
```